executor/calibrate_resource.go

```diff
@@ -211,34 +212,48 @@ func (e *calibrateResourceExec) dynamicCalibrate(ctx context.Context, req *chunk
 	if err != nil {
 		return err
 	}
-	rus, err := getRUPerSec(ctx, exec, startTime, endTime)
+	rus, err := getRUPerSec(ctx, e.ctx, exec, startTime, endTime)
 	if err != nil {
 		return err
 	}
-	tikvCPUs, err := getComponentCPUUsagePerSec(ctx, exec, "tikv", startTime, endTime)
+	tikvCPUs, err := getComponentCPUUsagePerSec(ctx, e.ctx, exec, "tikv", startTime, endTime)
 	if err != nil {
 		return err
 	}
-	tidbCPUs, err := getComponentCPUUsagePerSec(ctx, exec, "tidb", startTime, endTime)
+	tidbCPUs, err := getComponentCPUUsagePerSec(ctx, e.ctx, exec, "tidb", startTime, endTime)
 	if err != nil {
 		return err
 	}
 	quotas := make([]float64, 0)
 	lowCount := 0
-	for idx, ru := range rus {
-		if idx >= len(tikvCPUs) || idx >= len(tidbCPUs) {
+	for {
+		if rus.isEnd() || tikvCPUs.isEnd() || tidbCPUs.isEnd() {
 			break
 		}
-		tikvQuota, tidbQuota := tikvCPUs[idx]/totalKVCPUQuota, tidbCPUs[idx]/totalTiDBCPU
+		// make time point match
+		minTime := rus.getTime()
+		if tikvCPUs.getTime().Before(minTime) {
+			minTime = tikvCPUs.getTime()
+		}
+		if tidbCPUs.getTime().Before(minTime) {
+			minTime = tidbCPUs.getTime()
+		}
+		if !rus.findTime(minTime) || !tikvCPUs.findTime(minTime) || !tidbCPUs.findTime(minTime) {
```

Comment:
I think we should not break here, consider the following case(each number represent a timestamp):
ru ts: 9, 8, 7,  4, 3, 2, 1 ,0
kv cpu ts: 8, 7, 6, 5, 4, 3, 2, 1
db cpu: 5,4,3,2,1,0
After the first step, the minTime should be 5, then rus.findTime will point at 4 and thus return false, the iterator should not break but continue.

BTW: I perfer to let the iterator to use asc order because it's more intuitive.

Reply:
IMO, rus.findTime will point at 4 but return true

Reply:
This is also not correct because the other 2 point to 5

Reply:
I mean, https://github.com/pingcap/tidb/pull/43248#discussion_r1174757866 is useful.
THX!

Reply:
`ru3 and cpu3` will test this case.